src/sage/combinat/SJT.py

```diff
@@ -0,0 +1,141 @@
+            self.__directions = [-1] * self.__n
+
+            # The first element has null direction.
+            self.__directions[0] = 0
+        else:
+            self.__directions = directions
+
+    def __idx_largest_element_non_zero_direction(self, perm, directions):
+        r"""
+        Find the largest element in ``perm`` with a non null direction.
+        """
+        largest = 0
+        index = None
+        for i in range(self.__n):
+            if directions[i] != 0:
+                e = perm[i]
+                if e > largest:
+                    index = i
+                    largest = e
+
+        return index
+
+    def next(self):
+        r"""
+        Produce the next permutation following the Steinhaus-Johnson-Trotter
+        algorithm.
+
+        OUTPUT: the list of the next permutation.
+        """
+        # Copying lists of permutation and directions to avoid changing internal
+        # state of the algorithm if ``next()`` is called without reassigning.
+        perm = self.__perm[:]
+        directions = self.__directions[:]
+
+        # Assume that the element to move is n (which will be in most cases).
+        selected_elt = self.__n
+        xi = perm.index(selected_elt)
+        direction = directions[xi]
+
+        # If this element has null direction, find the largest whose is
+        # non-null.
+        if direction == 0:
+            xi = self.__idx_largest_element_non_zero_direction(perm, directions)
+            if xi is None:
+                # We have created every permutation. Detected when all elements
+                # have null direction.
+                return False, None
+            direction = directions[xi]
+            selected_elt = perm[xi]
+
+        new_pos = xi + direction
+
+        # Proceed to transpose elements and corresponding directions.
+        perm[xi], perm[new_pos] = perm[new_pos], perm[xi]
+        directions[xi], directions[new_pos] = \
+            directions[new_pos], directions[xi]
+
+        # If the transposition results in the largest element being on one edge
+        # or if the following element in its direction is greater than it, then
+        # then set its direction to 0
+        if new_pos == 0 or new_pos == self.__n - 1 or \
+            perm[new_pos + direction] > selected_elt:
+            directions[new_pos] = 0
+
+        # After each permutation, update each element's direction. If one
+        # element is greater than selected element, change its direction towards
+        # the selected element. This loops has no reason to be if selected
+        # element is n and this will be the case most of the time.
+        if selected_elt != self.__n:
+            for i in range(self.__n):
+                if perm[i] > selected_elt:
+                    if i < new_pos:
+                        directions[i] = 1
+                    if i > new_pos:
+                        directions[i] = -1
+
+        return perm, directions
```

Comment:
Is it a good idea to return another `SJT` element directly, or would the construction be slow?

Reply:
I've tried to do this. However, since a `SJT` object is created on each `Permutation` object creation, and because the check on the directions (if non-id perm is given and no directions: raises an error) is done on `SJT` creation, I don't think it's too bad as it is now.
`SJT` class is not meant to be used as is but more a support for `Permutation` to use.

What do you think ?